exp/runtime/hooks/api/v1alpha1/discovery_types.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+// DiscoveryRequest represents the object of a discovery request.
+// +kubebuilder:object:root=true
+type DiscoveryRequest struct {
+	metav1.TypeMeta `json:",inline"`
+}
+
+// DiscoveryResponse represents the object received as a discovery response.
```

Comment:
Note for a follow-up iteration. When we will start generating the OpenAPI spec we should revise those comments and make sure they are self-explanatory for people implementing extensions.
e.g. in this case, the extension implementer is required to return information about all the supported extensions etc.